File: examples/old/app/wgpu1_main.py
import numpy  # noqa
import glm

import kalast


def closure_update_state(
    state: kalast.gpu.win.StateStep,
) -> kalast.gpu.win.StateStep | None:
    if state.iteration == 0:
        rot = glm.mat3(
            glm.rotate(
                -90.0 * kalast.util.RPD,
                [1.0, 0.0, 0.0],
            )
        )
    else:
        rot = glm.mat3(
            glm.rotate(
                -100.0 * kalast.util.RPD * state.dt.total_seconds(), [0.0, 1.0, 0.0]
            )
        )

    # state.light_pos = state.light_pos @ rot

    for iib in range(0, len(state.models_state)):
        model = state.get_model_state(iib)
        # model.p[0] = 1.0

        model.m = model.m @ rot

        state.set_model_state(iib, model)

    return state


config = kalast.gpu.config.Config()
config.debug_state_creation = False
config.debug_state_rendering = False
config.width = 1440
config.height = 1080
config.depthpass_enable = False
config.render_light = True
config.hdr_enable = False
config.background = [0.0, 0.0, 0.0, 1.0]
config.texts = [
    # kalast.gpu.config.ConfigText("Hello custom text", pos=[1440.0, 0.0], ha="right"),
]
config.show_text_info = True
config.fps_time_refresh = 0.05

config.camera_pos = [0.0, 5.0, 10.0]
config.camera_yaw = -90.0
config.camera_pitch = -20.0
config.light_pos = [0.0, 0.0, 4.0]

# config.camera_pos = [0.0, -10.0, 0.0]
# config.camera_yaw = 90.0
# config.camera_pitch = 90.0
# config.light_pos = [0.0, -4.0, 0.0]

config.camera_fovy = 45.0
config.camera_znear = 0.1
config.camera_zfar = 100.0
config.camera_speed = 4.0
config.camera_sensitivity = 0.4
config.light_color = [1.0, 1.0, 1.0]

config.start_paused = True
config.global_test = 0
config.ambient_strength = 0.01
config.diffuse_enable = True
config.specular_enable = False

config.models = [
    kalast.gpu.config.ConfigModel(
        # path="/Users/gregoireh/data/mesh/didymos/didymos_g_9309mm_spc_obj_0000n00000_v003_decimate_1k.obj",
        path="/Users/gregoireh/data/mesh/didymos/didymos_g_9309mm_spc_obj_0000n00000_v003_decimated_3072.obj",
        # path="/Users/gregoireh/data/mesh/didymos/didymos_g_9309mm_spc_obj_0000n00000_v003.obj",
        # path="/Users/gregoireh/data/mesh/didymos/didymos_g_1165mm_spc_obj_0000n00000_v003.obj",
        flat=True,
        # pos_factor=[1.0, 1.0, 1.0],
        # color_mode=1,
        # color=[1.0, 1.0, 1.0]
    ),
]

kalast.gpu.win.run(config, closure_update_state)

File: examples/old/app/wgpu1_main.rs
use kalast;
use kalast::Vec3;

fn main() {
    // wgpu event logger
    env_logger::init();
    let ev = winit::event_loop::EventLoop::new().unwrap();
    ev.set_control_flow(winit::event_loop::ControlFlow::Poll);

    // let sun_au = 1.0;
    let sun_dir = Vec3::new(1.0, 0.0, 1.0).normalize();
    let light_dist = 10.0;

    let config = kalast::gpu::config::Config {
        width: 1440,
        height: 1080,
        render_light: true,
        enable_back_face: true,
        camera_pos: Vec3::new(5.0, 0.0, 0.0),
        camera_dir: Vec3::new(-1.0, 0.0, 0.0),
        camera_up: Vec3::new(0.0, 0.0, 1.0),
        camera_fovy: 10.0,
        camera_znear: 0.1,
        camera_zfar: 100.0,
        camera_speed: 4.0,
        camera_sensitivity: 0.4,
        light_pos: sun_dir * light_dist,
        light_color: Vec3::new(1.0, 1.0, 1.0),
        start_paused: true,

        ambient_strength: 0.5,

        // Not working?? looks like specular result instead of diffuse
        diffuse_enable: false,

        // Not working?? does nothing
        specular_enable: false,

        models: vec![kalast::gpu::config::ConfigModel {
            path: "res/plane_crater_1024-5000_h=0.437.obj".to_string(),
            flat: true,
            ..Default::default()
        }],
        ..Default::default()
    };

    let mut app = kalast::gpu::win::App::new(config);

    // app.run_blocked(ev);
    ev.run_app(&mut app).unwrap();
}
